### PR TITLE
feat(catalog): add pg_type.typinput and support dummy casting to regproc

### LINKED ISSUE #11689: bug: missing `typinput` in `pg_catalog.pg_type`
### Describe the bug

The following query was sent by DataGrip:

```sql
SELECT
    typinput = 'pg_catalog.array_in' :: regproc AS is_array,
    typtype,
    typname,
    pg_type.oid
FROM
    pg_catalog.pg_type
    LEFT JOIN (
        SELECT
            ns.oid AS nspoid,
            ns.nspname,
            r.r
        FROM
            pg_namespace AS ns
            JOIN (
                SELECT
                    s.r,
                    (current_schemas(false)) [s.r] AS nspname
                FROM
                    generate_series(1, array_upper(current_schemas(false), 1)) AS s(r)
            ) AS r USING (nspname)
    ) AS sp ON sp.nspoid = typnamespace
WHERE
    pg_type.oid = 1
ORDER BY
    sp.r,
    pg_type.oid DESC
```

### Error message/log

```text
ERROR:  QueryError: Bind error: failed to bind expression: typinput = CAST('pg_catalog.array_in' AS regproc)

Caused by:
  Item not found: Invalid column: typinput
```

### To Reproduce

_No response_

### Expected behavior

_No response_

### How did you deploy RisingWave?

_No response_

### The version of RisingWave

_No response_

### Additional context

https://www.postgresql.org/docs/current/catalog-pg-type.html

### PR DESCRIPTION
**Pull request by TennyZhuang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Fix #11689 

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: e2e_test/batch/catalog/pg_type.slt.part [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
+25	text	0	f	b text_in
+700	float4	0	f	b float4in
+701	float8	0	f	b float8in
+1043	varchar	0	f	b varchar_in
+1082	date	0	f	b date_in
+1083	time	0	f	b time_in
+1114	timestamp	0	f	b timestamp_in
+1184	timestamptz	0	f	b timestamptz_in
+1186	interval	0	f	b interval_in
+1301	rw_int256	0	f	b rw_int1256_in
```
Comment (yezizp2012) [inline e2e_test/batch/catalog/pg_type.slt.part:18]:
```suggestion
1301	rw_int256	0	f	b rw_int256_in
```

File: src/frontend/src/catalog/system_catalog/rw_catalog/rw_types.rs [hunk 2]
```diff
@@ -24,36 +24,41 @@ use risingwave_common::types::{DataType, ScalarImpl};
...
             $(
-            ($oid, stringify!($name)),
+            ($oid, stringify!($name), stringify!($input)),
             )*
             // Note: rw doesn't support `text` type, returning it is just a workaround to be compatible
             // with PostgreSQL.
-            (25, "text"),
-            (1301, "rw_int256"),
+            (25, "text", "text_in"),
+            (1301, "rw_int256", "rw_int1256_in"),
```
Comment (yezizp2012) [inline src/frontend/src/catalog/system_catalog/rw_catalog/rw_types.rs:35]:
```suggestion
            (1301, "rw_int256", "rw_int256_in"),
```

**Review by yezizp2012 — COMMENTED** [review]
Rest LGTM

File: src/common/src/types/postgres_type.rs [hunk 3]
```diff
@@ -30,21 +30,21 @@ macro_rules! for_all_base_types {
...
-            { Decimal     | 1700 |     1231 | numeric     |     -1 }
-            { Jsonb       | 3802 |     3807 | jsonb       |     -1 }
+            { Boolean     |   16 |     1000 | bool        | boolin         |      1 }
+            { Bytea       |   17 |     1001 | bytea       | byteain        |     -1 }
+            { Int64       |   20 |     1016 | int8        | int8in         |      8 }
+            { Int16       |   21 |     1005 | int2        | int2in         |      2 }
+            { Int32       |   23 |     1007 | int4        | int4in         |      4 }
+            { Float32     |  700 |     1021 | float4      | float4in       |      4 }
+            { Float64     |  701 |     1022 | float8      | float8in       |      8 }
+            { Varchar     | 1043 |     1015 | varchar     | varchar_in     |     -1 }
```
Comment (xiangjinwu) [inline src/common/src/types/postgres_type.rs:40]: `varcharin`

Is this result not copied directly from postgresql output? The macro was written to accept psql output format (`|` separated) as input directly.

File: src/frontend/src/catalog/system_catalog/rw_catalog/rw_types.rs [hunk 4]
```diff
@@ -24,36 +24,41 @@ use risingwave_common::types::{DataType, ScalarImpl};
...
         &[
             $(
-            ($oid, stringify!($name)),
+            ($oid, stringify!($name), stringify!($input)),
             )*
             // Note: rw doesn't support `text` type, returning it is just a workaround to be compatible
             // with PostgreSQL.
-            (25, "text"),
-            (1301, "rw_int256"),
+            (25, "text", "text_in"),
```
Comment (xiangjinwu) [inline src/frontend/src/catalog/system_catalog/rw_catalog/rw_types.rs:34]: `textin`

File: src/sqlparser/src/ast/data_type.rs [hunk 5]
```diff
@@ -56,6 +56,8 @@ pub enum DataType {
     Interval,
     /// Regclass used in postgresql serial
     Regclass,
+    /// Regproc used in postgresql function
+    Regproc,
```
[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/sqlparser/src/ast/data_type.rs:60]: Fix in later PR

**Pull request merged**